### DISCUSSION TITLE
Create and upload new sources from a directory of FLACs

### DISCUSSION THREAD

**Discussion by RogueOneEcho** [RogueOneEcho/caesura#197] [Ideas]
`caesura` is designed for transcoding from sources that **already exist** on RED or OPS.

In order to upload an entirely new release there's a lot of metadata that needs to be collected and that's outside the scope of the tool at present. 

Metadata could be retrieve from sources such as MusicBrainz but quality of the data is uncertain.

If the community wishes to contribute an implementation then please discuss here but at present this isn't on the roadmap.

Related: #196

**Comment by tomerh2001** [comment]
I did three research passes (RED requirements, caesura architecture, and gazelle_api scope) and I think there’s a safe path to contribute this.

Proposed contribution split:

1. `gazelle_api` PR first  
- Add generic new-source upload support for `ajax.php?action=upload` (separate form + client method).  
- Keep it tracker-agnostic (no RED/OPS policy logic in this crate).

2. `caesura` PR second (after the `gazelle_api` release)  
- Add a new `publish` flow for local FLAC source upload, then optional V0/320 upload.  
- Keep existing `upload`/`batch` behavior unchanged.  
- Start RED-only with explicit manual safety gates (metadata confirmation, dupe/DNU checklist acknowledgement, edition/media correctness).

If this boundary/scope looks good to you, I’ll work on the `gazelle_api` PR first and then follow with the `caesura` PR pinned to the released crate version.

**Reply by RogueOneEcho** [reply]
> Keep it tracker-agnostic (no RED/OPS policy logic in this crate).
You're going to struggle with that. OPS and RED have a lot of subtle differences.

> Add a new publish flow for local FLAC source upload

There's some complexity you've missed 

1. Adding to an exist torrent group 
2. Creating a new torrent group 

Both will require different metadata, I'm not sure what the format is. 

https://github.com/OPSnet/Gazelle/blob/master/docs/07-API.md#upload

I take it you're leaving it to the user to fetch the meta data, then how do you pass it to caesura? cli args would likely be too lengthy so probably a yaml file? 


```
caesura publish [PATH TO YAML] 
```


> then optional V0/320 upload.

Simplest solution would be for the publish command to just publish the FLAC, then the normal transcode command can be run with the returned Id

> dupe

Dupe detection is going to be tricky. If publishing to an existing torrent group it might be worth using a fuzzy matching crate for checking the catelogue number is unique.

> If this boundary/scope looks good to you, I’ll work on the gazelle_api PR first and then follow with the caesura PR pinned to the released crate version.

Sounds fine

**Reply by tomerh2001** [reply]
Thanks, your notes were spot on.

I opened the `gazelle_api` PR here: https://github.com/RogueOneEcho/gazelle_api/pull/3

I narrowed PR1 to API-surface work and incorporated the key points you raised:
- separate `upload_new_source` flow (existing `upload_torrent` path unchanged)
- numeric `releasetype` mapping (not label-based)
- optional `requestid` support for request-fill uploads
- conditional remaster/unknown mapping (no forced tracker-policy behavior)
- trait/mock wiring + tests (including ignored integration test pattern)

I’m treating this as RED-first validated and not claiming full OPS/RED parity yet.

For the follow-up `caesura` PR, I’ll take the flow you suggested:
1. publish source FLAC first,
2. return/use the uploaded torrent ID,
3. run normal transcode/upload after that,
with YAML manifest input and explicit dupe checks for existing-group paths (including catalogue matching logic).

If this PR1 API shape looks good to you, I’ll proceed to work on the `caesura` PR on top of the released crate.

**Reply by tomerh2001** [reply]
Also opened caesura #198 with a first-cut implementation of local source (FLAC) publishing.

Assuming [PR1](https://github.com/RogueOneEcho/gazelle_api/pull/3) (gazelle_api) and [[#198](https://github.com/RogueOneEcho/caesura/pull/198)](this) PR get merged, my next step will be to improve metadata population (likely via MusicBrainz). I’ll also review how yadg-pth-userscript approaches metadata extraction to see what we can reuse or adapt (or even directly use).

As-is, this version gets the core flow working end-to-end: 
1. publish a new source torrent from local FLACs.
2. Then you can run the existing transcode/upload flow on the returned torrent ID.

**Reply by RogueOneEcho** [reply]
Thanks for contributing #198.

I've been going back and forth on this and have concluded that source publishing is beyond what caesura is intended for. 

caesura is focused on transcoding existing sources, and while your prototype is well executed and far more manageable than I had envisioned, it does complicate caesura and adds a significant future maintenance burden. 

Given there hasn't been much demand for this functionality, I think it would be better suited as a separate tool. 

If the separate tool reuses some of caesura's core functionality we could look at exposing that as a shared `pub` library.

I'm therefore going to close #198
